README.md
# flutter_refresh_loadmore

A refresh loadmore listview package.

## Getting Started
Installing
 ```dart
flutter_refresh_loadmore: ^0.0.9
```

Import
 ```dart
import 'package:flutter_refresh_loadmore/flutter_refresh_loadmore.dart';
```

How To Use

 ```dart
     bool hasLoadMore=true;

   //模拟刷新
   Future _refresh() async{
     await Future.delayed(Duration(milliseconds: 1000),(){
       length=15;
       hasLoadMore=true;
     });
     _globalKey.currentState.updateUi(length, hasLoadMore);

   }
   //模拟加载更多
   Future _loadmore() async {
       await Future.delayed(Duration(milliseconds: 1000),(){
         //模拟 没有更多数据
         if(length>=30){
           hasLoadMore=false;
         }else{
           //正常加载
           length=length+5;
         }

       });
       _globalKey.currentState.updateUi(length, hasLoadMore);
   }

   int length=15;
   GlobalKey<RefreshLoadMoreState> _globalKey=new GlobalKey();
   @override
   Widget build(BuildContext context) {
     return Scaffold(

       appBar: AppBar(
         centerTitle: true,
         title: Text("test"),
       ),
       body: Center(
         child: RefreshLoadMoreWidget(

           key: _globalKey,
           //必传  listview 的 item widget
           swrapInsideWidget: (context, index) {
             return Container(
               alignment: Alignment.center,
               margin: EdgeInsets.only(top: 10),
               color: Colors.deepOrange,
               child: Text("$index"),
               width: 100,
               height: 50,
             );
           },
           //必传 加载更多是否还有数据
           hasMoreData: hasLoadMore,

           //不传 默认没有上拉刷新功能 必须有async  await
           refrshCallback: () async {
            await _refresh();
           },
           //必传
           itemCount: length,

           //不传 默认没有加载更多功能     必须有async  await
           loadMoreCallback: () async {
             await _loadmore();
           },
 //          //不传有默认的  自定义自己的footer  str  值有两个  加载中...  和 暂无更多数据
 //          swrapFooterWidget:(str){
 //           return Center(
 //             child: Text(str),
 //           );
 //          } ,
 //          //不传有默认的 自定义自己的head  statu  值有4个  正加载中     加载完成 下拉刷新  松开刷新
 //          swrapHeadWidget: (statu){
 //            return SpinKitFadingCircle(color: Colors.black38,size: 30,);
 //          },

         ),
       ),
     );
   }
 ```


下拉默认刷新效果
https://github.com/liuwangle/flutter_refresh_loadmore/blob/master/demon_gif/refresh.gif



加载默认更多效果
https://github.com/liuwangle/flutter_refresh_loadmore/blob/master/demon_gif/loadmore.gif

CHANGELOG.md
## [0.0.1] - 2019/6/25

* 下拉刷新  上拉加载更多


test/flutter_refresh_loadmore_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_refresh_loadmore/flutter_refresh_loadmore.dart';

void main() {
  test('adds one to input values', () {

  });
}


lib/rotate_widget.dart
import 'package:flutter/material.dart';
import 'dart:math' as math;

class RotateWidget extends StatefulWidget {
  bool isStartAnimation;

  RotateWidget({this.isStartAnimation});

  @override
  State<StatefulWidget> createState() {
    return _RotateWidget();
  }
}

class _RotateWidget extends State<RotateWidget> with TickerProviderStateMixin {
  AnimationController controller;
  Animation _value;


  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }
  @override
  void initState() {
    controller = AnimationController(
        duration: Duration(milliseconds: 800), vsync: this);
    Tween tween = Tween(begin: 0, end: math.pi * 2);
    _value = tween.animate(controller);
    controller.addListener(() {
      if(controller.value==1){
        controller.value=0;
      }
      setState(() {});
    });
    controller.addStatusListener((status) {
      if (status == AnimationStatus.completed) {
//        controller.value = 0;
//        controller.forward();
      }
    });

    super.initState();
//    controller.forward();
//    Matrix4.rotate(Vector3(25,25,25), controller.value)
  }

  @override
  Widget build(BuildContext context) {
//    if(widget.mykey.currentContext.size!=null){
//      if(widget.mykey.currentContext.size.height>0&&!controller.isAnimating){
//
//      }
//    }
    if (!controller.isAnimating && widget.isStartAnimation) {
      controller.forward();
    } else if (!widget.isStartAnimation && controller.isAnimating) {
      controller.stop(canceled: true);
    }


    Transform transform=Transform.rotate(
      //旋转90度 pi/2
      angle: _getValue(),
      child: Image.asset(

        'loading.png',
        package:"flutter_refresh_loadmore",
        width: 25.0,
        height: 25.0,
        fit: BoxFit.fitWidth,

      ));
//    print("xxxxxxxxxxxxxxxxxx1 \n  ${transform.transform}");
//    print("xxxxxxxxxxxxxxxxxx2 \n  ${Matrix4.identity()..rotateZ(_getValue())}");
    var mwidget =

     Container(
       width: 25.0,
       height: 25.0,

      child:transform,

    );
//    Container(
//
//      alignment: Alignment.center,
//      width: 50.0,
//      height: 50.0,
//
//      child: Image.asset(
//        'images/loading_11.png',
//        width: 50.0,
//        height: 50.0,
//        fit: BoxFit.cover,
//      ),
//    ) ;

    return mwidget;
  }

  double _getValue() {
    return double.parse(_value.value.toString());
  }
}


lib/flutter_refresh_loadmore.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'dart:math' as math;

import 'rotate_widget.dart';
import 'refresh_head.dart';

typedef SwrapInsideWidget = Widget Function(BuildContext context, int index);

typedef RefrshCallback = Future Function();
typedef LoadMoreCallback = Future<void> Function();

class RefreshLoadMoreWidget extends StatefulWidget {
  //listview 的item  widget
  SwrapInsideWidget swrapInsideWidget;

  //headviw
  SwrapHeadWidget swrapHeadWidget;

  //刷新的回调  不传或者 传null  则关闭刷新
  RefrshCallback refrshCallback;

  //加载更多回调  不传或者 传null  则关闭加载更多
  LoadMoreCallback loadMoreCallback;
  bool hasMoreData = true;
  int itemCount;
  SwrapFooterWidget swrapFooterWidget;

  RefreshLoadMoreWidget(
      {Key key,
        @required this.swrapInsideWidget,
        this.refrshCallback,
        this.loadMoreCallback,
        this.swrapHeadWidget,
        this.swrapFooterWidget,
        @required this.hasMoreData,
        @required this.itemCount}):super(key:key);

  @override
  State<StatefulWidget> createState() {
    return RefreshLoadMoreState();
  }
}

class RefreshLoadMoreState extends State<RefreshLoadMoreWidget> {
  ScrollController controller = ScrollController();

  /**
   * realse  动画
   */
  AnimationController _positionController;

  @override
  void dispose() {
    if (_positionController != null) {
      _positionController.dispose();
    }

    super.dispose();
  }

  updateUi(int itemCount,bool hasMoredata){
    widget.itemCount=itemCount;
    widget.hasMoreData=hasMoredata;
    setState(() {

    });
  }

  //是否正在加载更多
  bool isLoadingMore = false;

  //刷新widget 高度
  double headHeight = 60;

  @override
  void initState() {
    controller.addListener(() {
      //总长  减去  当时的长度  35 是loadmore 的高度
      var offset =
          controller.position.maxScrollExtent - controller.position.pixels;

//      if (offset >= 0 &&
//          offset <= 35 &&
//          !isLoadingMore &&
//          widget.loadMoreCallback() != null &&
//          widget.hasMoreData) {
//        setState(() {
//          isLoadingMore = true;
//        });
//        widget.loadMoreCallback().whenComplete(() {
//          setState(() {
//            isLoadingMore = false;
//          });
//        });
//      }
    });
    WidgetsBinding widgetsBinding = WidgetsBinding.instance;
    //第一帧 绘制完毕
    widgetsBinding.addPostFrameCallback((callback) {});
    super.initState();
  }

  bool isRefreshing=false;

  //回调刷新  然后继续执行动画
  _refresh() async {
    if(isRefreshing){
      return;
    }
    isRefreshing=true;

    widget.refrshCallback().whenComplete(() {
      isRefreshing=false;
      _odelta = headHeight;
      _isLoadingHead = false;
      _positionController.value = 0;
      _positionController.duration = Duration(milliseconds: 300);
      _positionController.forward();
    });
  }

  static double _odelta = 0;

  bool _handleScrollNotification(ScrollNotification notification) {
    if(isRefreshing){
      return true;
    }
    //listview 列表滑动
    if (notification is ScrollUpdateNotification) {}
    //listview  滑动到尽头 继续滑动
    if (notification is OverscrollNotification) {
      if (!_positionController.isAnimating) {
        _odelta -= notification.overscroll / 2;
        if (notification.metrics.pixels == 0) {
          if (headKey.currentState != null) {

            //head 高度变化
            headKey.currentState.setState(() {});
          }
        }
      }
    }
    //end 滑动结束
    if (notification is ScrollEndNotification) {}

    //滑动状态改变
    if (notification is UserScrollNotification) {
      //滑动空闲状态 或者 方向状态 执行动画（手指抬起 或者 反向）
      if ((notification.direction == ScrollDirection.idle && _odelta > 0) ||
          (notification.direction == ScrollDirection.reverse &&
              notification.metrics.pixels == 0 &&
              _odelta > 0)) {
        _isUseAnimationValue = true;

        if (_odelta > headHeight) {
          _isLoadingHead = true;
        } else {
          _isLoadingHead = false;
        }

        if (_positionController.isAnimating) {
          _odelta = _odelta * _positionController.value;
          _positionController.stop(canceled: true);
        } else {
          _positionController.value = 0;
        }

        _positionController.duration = Duration(milliseconds: 400);
        _positionController.forward();
      }
    }

    //如果headview 显示   listview 不可滑动
    if (_odelta > 0) {
      if (_scrollPhysics != const RefreshScrollPhysics()) {
        _scrollPhysics = const RefreshScrollPhysics();
        setState(() {});
      }
    } else {
      if (_scrollPhysics != const RefreshAlwaysScrollPhysics()) {
        _scrollPhysics = const RefreshAlwaysScrollPhysics();
        setState(() {});
      }
      ;
    }
  }

  bool _handleGlowNotification(OverscrollIndicatorNotification notification) {
    return true;
  }

  //headwidget 是否正在加载中
  static bool _isLoadingHead = false;

  //head view  的高度是否 用 动画的值
  bool _isUseAnimationValue = false;

  //获取head widget  高度
  double _getHeight() {
    //正在执行动画的时候 head widget的高度
    if (_isUseAnimationValue) {
      double h = (1 - _positionController.value) * _odelta;
      if (_isLoadingHead) {
        h = (1 - _positionController.value) * (_odelta - headHeight) +
            headHeight;
      } else {
        h = (1 - _positionController.value) * _odelta;
      }

      return h;
    }
    if (_odelta < 0) {
      return 0;
    }
    return _odelta;
  }

  Widget buildLoadMore() {

    if(widget.swrapFooterWidget!=null){
      return widget.swrapFooterWidget(widget.hasMoreData?"加载中...":"暂无更多数据");
    }
    return Padding(
      padding: EdgeInsets.all(10),
      child: Align(
          alignment: Alignment.center,
          child: widget.hasMoreData
              ? Row(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              RotateWidget(
                isStartAnimation: true,
              ),
              SizedBox(
                width: 15,
              ),
              Text(
                "加载中...",
                style: TextStyle(fontSize: 18, color: Color(0Xff222222)),
              )
            ],
          )
              : Text(
            "暂无更多数据",
            style: TextStyle(color: Color(0xff999999)),
          )),
    );
  }

  //listview  控制是否可滑动
  ScrollPhysics _scrollPhysics = const RefreshAlwaysScrollPhysics();

  //Headwidget
  DefaultHead defaultHead;
  GlobalKey<State> listKey = new GlobalKey();
  GlobalKey<State> notifKey = new GlobalKey();

  GlobalKey<State> headKey = new GlobalKey();

  @override
  Widget build(BuildContext context) {
    defaultHead = DefaultHead(
      key: headKey,
      heightsCallBack: _getHeight,
      getAnimationController: (controller) {
        _positionController = controller;
      },
      oneceAnimationComplete: () {
        _odelta = headHeight;
        _refresh();
      },
      headAnimationComplete: () {
        //最后的动画
        _odelta = 0;
        _isUseAnimationValue = false;
        _scrollPhysics = const RefreshAlwaysScrollPhysics();
        setState(() {});
      },
      normalHeight: headHeight,
      swrapHeadWidget: widget.swrapHeadWidget,
    );

    ListView mylistview = ListView.builder(
      physics: _scrollPhysics,
//      padding: const EdgeInsets.all(8.0),
      controller: controller,
      itemCount: widget.loadMoreCallback == null
          ? widget.itemCount
          : widget.itemCount + 1,
      itemBuilder: (BuildContext context, int index) {
        if (index == (widget.itemCount) && widget.loadMoreCallback != null) {
          if ((!isLoadingMore )&&(widget.loadMoreCallback != null )&&(widget.hasMoreData)) {
            isLoadingMore = true;
            widget.loadMoreCallback().whenComplete(() {
              isLoadingMore = false;
            });
          }

          return buildLoadMore();
        }
        return widget.swrapInsideWidget(context, index);
      },
    );
    Widget mynotification = NotificationListener<ScrollNotification>(
        key: notifKey,
        onNotification: _handleScrollNotification,
        child: NotificationListener<OverscrollIndicatorNotification>(
          onNotification: _handleGlowNotification,
          child: ScrollConfiguration(
            behavior: MyBehavior(false, false, Colors.red),
            child: mylistview,
          ),
        ));

    return Column(
      children: <Widget>[
        widget.refrshCallback == null ? Container() : defaultHead,
        new Expanded(
          child: mynotification,
          flex: 1,
        )
      ],
    );
  }
}

///切记 继承ScrollPhysics  必须重写applyTo，，在NeverScrollableScrollPhysics类里面复制就可以
///出现反向滑动时用此ScrollPhysics
class RefreshScrollPhysics extends ScrollPhysics {
  const RefreshScrollPhysics({ScrollPhysics parent}) : super(parent: parent);

  @override
  RefreshScrollPhysics applyTo(ScrollPhysics ancestor) {
    return new RefreshScrollPhysics(parent: buildParent(ancestor));
  }

  @override
  bool shouldAcceptUserOffset(ScrollMetrics position) {
    return true;
  }

  ///防止ios设备上出现弹簧效果
  @override
  double applyBoundaryConditions(ScrollMetrics position, double value) {
    assert(() {
      if (value == position.pixels) {
        throw FlutterError(
            '$runtimeType.applyBoundaryConditions() was called redundantly.\n'
                'The proposed new position, $value, is exactly equal to the current position of the '
                'given ${position.runtimeType}, ${position.pixels}.\n'
                'The applyBoundaryConditions method should only be called when the value is '
                'going to actually change the pixels, otherwise it is redundant.\n'
                'The physics object in question was:\n'
                '  $this\n'
                'The position object in question was:\n'
                '  $position\n');
      }
      return true;
    }());
    if (value < position.pixels &&
        position.pixels <= position.minScrollExtent) // underscroll
      return value - position.pixels;
    if (position.maxScrollExtent <= position.pixels &&
        position.pixels < value) // overscroll
      return value - position.pixels;
    if (value < position.minScrollExtent &&
        position.minScrollExtent < position.pixels) // hit top edge
      return value - position.minScrollExtent;
    if (position.pixels < position.maxScrollExtent &&
        position.maxScrollExtent < value) // hit bottom edge
      return value - position.maxScrollExtent;
    return 0.0;
  }

  //重写这个方法为了减缓ListView滑动速度
  @override
  double applyPhysicsToUserOffset(ScrollMetrics position, double offset) {
    if (offset < 0.0) {
      return 0.00000000000001;
    }
    if (offset == 0.0) {
      return 0.0;
    }
    return offset / 2;
  }

  //此处返回null时为了取消惯性滑动
  @override
  Simulation createBallisticSimulation(
      ScrollMetrics position, double velocity) {
    return null;
  }
}

///可去掉过度滑动时ListView顶部的蓝色光晕效果
class MyBehavior extends ScrollBehavior {
  final bool isShowLeadingGlow;
  final bool isShowTrailingGlow;
  final Color _kDefaultGlowColor;

  MyBehavior(
      this.isShowLeadingGlow, this.isShowTrailingGlow, this._kDefaultGlowColor);

  @override
  Widget buildViewportChrome(
      BuildContext context, Widget child, AxisDirection axisDirection) {
    //如果头部或底部有一个 不需要 显示光晕时 返回GlowingOverscrollIndicator
    if (!isShowLeadingGlow || !isShowTrailingGlow) {
      return new GlowingOverscrollIndicator(
        showLeading: isShowLeadingGlow,
        showTrailing: isShowTrailingGlow,
        child: child,
        axisDirection: axisDirection,
        color: _kDefaultGlowColor,
      );
    } else {
      //都需要光晕时  返回系统默认
      return super.buildViewportChrome(context, child, axisDirection);
    }
  }
}

///切记 继承ScrollPhysics  必须重写applyTo，，在NeverScrollableScrollPhysics类里面复制就可以
///此类用来控制IOS过度滑动出现弹簧效果
class RefreshAlwaysScrollPhysics extends AlwaysScrollableScrollPhysics {
  const RefreshAlwaysScrollPhysics({ScrollPhysics parent})
      : super(parent: parent);

  @override
  RefreshAlwaysScrollPhysics applyTo(ScrollPhysics ancestor) {
    return new RefreshAlwaysScrollPhysics(parent: buildParent(ancestor));
  }

  @override
  bool shouldAcceptUserOffset(ScrollMetrics position) {
    return true;
  }

  ///防止ios设备上出现弹簧效果
  @override
  double applyBoundaryConditions(ScrollMetrics position, double value) {
    assert(() {
      if (value == position.pixels) {
        throw FlutterError(
            '$runtimeType.applyBoundaryConditions() was called redundantly.\n'
                'The proposed new position, $value, is exactly equal to the current position of the '
                'given ${position.runtimeType}, ${position.pixels}.\n'
                'The applyBoundaryConditions method should only be called when the value is '
                'going to actually change the pixels, otherwise it is redundant.\n'
                'The physics object in question was:\n'
                '  $this\n'
                'The position object in question was:\n'
                '  $position\n');
      }
      return true;
    }());
    if (value < position.pixels &&
        position.pixels <= position.minScrollExtent) // underscroll
      return value - position.pixels;
    if (position.maxScrollExtent <= position.pixels &&
        position.pixels < value) // overscroll
      return value - position.pixels;
    if (value < position.minScrollExtent &&
        position.minScrollExtent < position.pixels) // hit top edge
      return value - position.minScrollExtent;
    if (position.pixels < position.maxScrollExtent &&
        position.maxScrollExtent < value) // hit bottom edge
      return value - position.maxScrollExtent;
    return 0.0;
  }

  ///防止ios设备出现卡顿
  @override
  Simulation createBallisticSimulation(
      ScrollMetrics position, double velocity) {
    final Tolerance tolerance = this.tolerance;
    if (position.outOfRange) {
      double end;
      if (position.pixels > position.maxScrollExtent)
        end = position.maxScrollExtent;
      if (position.pixels < position.minScrollExtent)
        end = position.minScrollExtent;
      assert(end != null);
      return ScrollSpringSimulation(spring, position.pixels,
          position.maxScrollExtent, math.min(0.0, velocity),
          tolerance: tolerance);
    }
    if (velocity.abs() < tolerance.velocity) return null;
    if (velocity > 0.0 && position.pixels >= position.maxScrollExtent)
      return null;
    if (velocity < 0.0 && position.pixels <= position.minScrollExtent)
      return null;
    return ClampingScrollSimulation(
      position: position.pixels,
      velocity: velocity,
      tolerance: tolerance,
    );
  }
}


lib/refresh_head.dart
import 'package:flutter/material.dart';
import 'dart:math' as math;

import 'rotate_widget.dart';

//高度回调
typedef HeightsCallBack = double Function();

//animationcontroller 回调
typedef GetAnimationController = void Function(AnimationController controller);

//head  动画执行完成回调
typedef DefaultHeadAnimationComplete = void Function();

typedef SwrapHeadWidget = Widget Function(String statusStr);
typedef SwrapFooterWidget = Widget Function(String statusStr);

class DefaultHead extends StatefulWidget {
  HeightsCallBack heightsCallBack;
  GetAnimationController getAnimationController;
  DefaultHeadAnimationComplete oneceAnimationComplete;
  DefaultHeadAnimationComplete headAnimationComplete;

  double normalHeight;
  GlobalKey key;
  SwrapHeadWidget swrapHeadWidget;

  DefaultHead(
      {this.key,
      this.heightsCallBack,
      this.getAnimationController,
      this.oneceAnimationComplete,
      this.headAnimationComplete,
      this.swrapHeadWidget,
      this.normalHeight})
      : super(key: key);

  _DefaultHeadState __defaultHeadState;

  @override
  State<StatefulWidget> createState() {
    if (__defaultHeadState == null) {
      __defaultHeadState = _DefaultHeadState();
    }

    return __defaultHeadState;
  }
}

class _DefaultHeadState extends State<DefaultHead>
    with TickerProviderStateMixin {
  AnimationController positionController;

//  AnimationController controller;
//  CurvedAnimation curved;
//  Animation  rotate;

  _DefaultHeadState();

  @override
  void initState() {
//    controller = AnimationController( duration: const Duration(milliseconds: 1000),vsync: this);
    positionController = AnimationController(
        duration: const Duration(milliseconds: 400), vsync: this);

//    curved = new CurvedAnimation(parent: controller, curve:Interval(0.1, 0.3,curve:  Curves.linear),);
//    rotate=Tween<double>(begin: 0, end: 10,).animate(controller);
    positionController.addListener(() {
      setState(() {});
    });
    positionController.addStatusListener((status) {
      if (AnimationStatus.completed == status) {
        if (widget.heightsCallBack() > 0) {
          widget.oneceAnimationComplete();
        } else {
          widget.headAnimationComplete();
          lastStatu = "";
        }
      }
    });
    widget.getAnimationController(positionController);

    super.initState();
//    controller.repeat();
  }

  RotateWidget rotateWidget;

  bool shouldIsRotate = false;

  @override
  Widget build(BuildContext context) {
    rotateWidget = RotateWidget(
      isStartAnimation: shouldIsRotate,
    );
    print("fffffffffffffff  ${shouldIsRotate}");
    return Container(
      color: Colors.black12,
      height: widget.heightsCallBack(),
      child: widget.swrapHeadWidget == null
          ? Container(
              margin: EdgeInsets.only(bottom: 20),
              height: widget.normalHeight,
              alignment: Alignment.bottomCenter,
              child: Row(
                mainAxisAlignment: MainAxisAlignment.center,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: <Widget>[
                  rotateWidget,
                  SizedBox(
                    width: 10,
                  ),
                  Text(
                    getText(),
                    style: TextStyle(fontSize: 15),
                  )
                ],
              ))
          : widget.swrapHeadWidget(getText()),
    );
  }

  String lastStatu = "";

  String getText() {
    String statu = "";
    if (widget.heightsCallBack() > widget.normalHeight) {
      if (positionController.isAnimating) {
        statu = "正加载中";
      } else {
        statu = "松开刷新";
      }
    } else if (widget.heightsCallBack() == widget.normalHeight) {
      statu = "正加载中";
    } else {
      if (lastStatu == "加载完成" || lastStatu == "正加载中") {
        statu = "加载完成";
      } else {
        statu = "下拉刷新";
      }
    }

    lastStatu = statu;
    if ("正加载中" == statu) {
      shouldIsRotate = true;
    } else {
      shouldIsRotate = false;
    }
    return statu;
  }
}
